packages/fx-core/src/component/drivers/aad/create.ts

```diff
@@ -0,0 +1,118 @@
+import { CreateAadAppOutput } from "./interface/createAadAppOutput";
+import { M365TokenProvider, UserError } from "@microsoft/teamsfx-api";
+import { GraphScopes } from "../../../common/tools";
+import { Constants } from "../../resource/aadApp/constants";
+import * as DriverConstants from "./constants";
+import { getDefaultString, getLocalizedString } from "../../../common/localizeUtils";
+
+const actionName = "aadApp/create"; // DO NOT MODIFY the name
+const helpLink = "https://aka.ms/teamsfx-actions/aadapp-create";
+
+@Service(actionName) // DO NOT MODIFY the service name
+export class CreateAadAppDriver implements StepDriver {
+  public async run(args: CreateAadAppArgs, context: DriverContext): Promise<Map<string, string>> {
+    this.validateArgs(args);
+    const aadAppClient = new AadAppClient(context.m365TokenProvider);
+    const aadAppState = this.loadCurrentState();
+
+    if (!aadAppState.AAD_APP_CLIENT_ID) {
+      // Create new AAD app if no client id exists
+      const aadApp = await aadAppClient.createAadApp(args.name);
+      aadAppState.AAD_APP_CLIENT_ID = aadApp.appId!;
+      aadAppState.AAD_APP_OBJECT_ID = aadApp.id!;
+      this.setAadEndpointInfo(context.m365TokenProvider, aadAppState);
+    }
+
+    if (args.genegerateClientSecret && !aadAppState.SECRET_AAD_APP_CLIENT_SECRET) {
+      // Create new client secret if no client secret exists
+      if (!aadAppState.AAD_APP_OBJECT_ID) {
+        throw new UserError({
+          source: actionName,
+          name: DriverConstants.missingEnvError.errorCode,
+          message: getDefaultString(
+            DriverConstants.missingEnvError.messageKey,
+            getDefaultString(DriverConstants.generateSecretError.messageKey),
+            "AAD_APP_OBJECT_ID"
+          ),
+          displayMessage: getLocalizedString(
+            DriverConstants.missingEnvError.messageKey,
+            getLocalizedString(DriverConstants.generateSecretError.messageKey),
+            "AAD_APP_OBJECT_ID"
+          ),
+          helpLink: helpLink,
+        });
+      }
+      const secret = await aadAppClient.generateClientSecret(aadAppState.AAD_APP_OBJECT_ID);
+      aadAppState.SECRET_AAD_APP_CLIENT_SECRET = secret;
+    }
+
+    return new Map(
+      Object.entries(aadAppState) // convert each property to Map item
+        .filter((item) => item[1] && item[1] !== "") // do not return Map item that is empty
+    );
+  }
+
+  private validateArgs(args: CreateAadAppArgs): void {
+    const invalidParameters: string[] = [];
+    if (!args.name) {
+      invalidParameters.push("name");
+    }
+
+    if (args.genegerateClientSecret === undefined) {
+      invalidParameters.push("genegerateClientSecret");
+    }
+
+    if (invalidParameters.length > 0) {
+      const invalidParametersString = invalidParameters.join(",");
+      throw new UserError({
+        source: actionName,
+        name: DriverConstants.invalidParameterError.errorCode,
+        message: getDefaultString(
```

Comment:
maybe you can throw PrerequisiteError here, the code is more simple than construct a UserError

Reply:
Thanks. I found I need some non-trivial changes to PrerequisiteError in order to adopt it, so followed your idea to create my own error class. We can see how to share the common parts between each action (driver) later.